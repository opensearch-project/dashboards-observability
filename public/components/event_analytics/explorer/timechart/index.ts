/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { HitsCounter } from './hits_counter/hits_counter';
export { Timechart } from './timechart';
export { getTimeRangeFromCountDistribution } from './utils';
