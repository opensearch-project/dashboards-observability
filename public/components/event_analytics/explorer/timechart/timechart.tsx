/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiSpacer } from '@elastic/eui';
import _ from 'lodash';
import React from 'react';
import { selectCountDistribution } from '../../redux/slices/count_distribution_slice';
import { CountDistribution } from '../visualizations/count_distribution';
import { HitsCounter } from './hits_counter';
import { TimechartHeader } from './timechart_header';

interface TimechartProps {
  countDistribution: ReturnType<typeof selectCountDistribution>[string];
  timeIntervalOptions: Array<{ text: string; value: string }>;
  onChangeInterval: (interval: string) => void;
  selectedInterval: string;
  startTime?: string;
  endTime?: string;
}

export const Timechart: React.FC<TimechartProps> = (props) => {
  return (
    <>
      <HitsCounter
        hits={_.sum(props.countDistribution.data?.['count()'])}
        showResetButton={false}
        onResetQuery={() => {}}
      />
      <TimechartHeader
        options={props.timeIntervalOptions}
        onChangeInterval={props.onChangeInterval}
        stateInterval={props.selectedInterval}
        startTime={props.startTime}
        endTime={props.endTime}
      />
      <EuiSpacer size="s" />
      <CountDistribution
        countDistribution={props.countDistribution}
        selectedInterval={props.selectedInterval}
        startTime={props.startTime}
        endTime={props.endTime}
      />
    </>
  );
};
