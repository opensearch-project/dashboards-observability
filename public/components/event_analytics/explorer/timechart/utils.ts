/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { selectCountDistribution } from '../../redux/slices/count_distribution_slice';

export const getTimeRangeFromCountDistribution = (
  countDistribution: ReturnType<typeof selectCountDistribution>[string]
): { startTime?: string; endTime?: string } => {
  const {
    data,
    metadata: { fields },
  } = countDistribution;
  // fields[1] is the x-axis (time buckets) in count distribution
  return { startTime: data[fields[1].name].at(0), endTime: data[fields[1].name].at(-1) };
};
