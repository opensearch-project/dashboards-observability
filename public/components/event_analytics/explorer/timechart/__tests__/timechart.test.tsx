/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render } from '@testing-library/react';
import React, { ComponentProps } from 'react';
import { Timechart } from '../timechart';

const renderTimechart = (overrideProps: Partial<ComponentProps<typeof Timechart>> = {}) => {
  const props: jest.Mocked<ComponentProps<typeof Timechart>> = Object.assign<
    ComponentProps<typeof Timechart>,
    Partial<ComponentProps<typeof Timechart>>
  >(
    {
      countDistribution: {
        selectedInterval: 'y',
        data: { 'count()': [194], 'span(timestamp,1y)': ['2024-01-01 00:00:00'] },
        metadata: {
          fields: [
            { name: 'count()', type: 'integer' },
            { name: 'span(timestamp,1y)', type: 'timestamp' },
          ],
        },
        size: 1,
        status: 200,
        jsonData: [{ 'count()': 194, 'span(timestamp,1y)': '2024-01-01 00:00:00' }],
      },
      timeIntervalOptions: [
        { text: 'Minute', value: 'm' },
        { text: 'Hour', value: 'h' },
        { text: 'Day', value: 'd' },
        { text: 'Week', value: 'w' },
        { text: 'Month', value: 'M' },
        { text: 'Year', value: 'y' },
      ],
      onChangeInterval: jest.fn(),
      selectedInterval: 'y',
      startTime: '2024-01-01 00:00:00',
      endTime: '2024-01-01 00:00:00',
    },
    overrideProps
  );
  const component = render(<Timechart {...props} />);
  return { component, props };
};

describe('<Timechart /> spec', () => {
  it('should change to week', () => {
    const { component, props } = renderTimechart();
    fireEvent.change(component.getByTestId('eventAnalytics__EventIntervalSelect'), {
      target: { value: 'w' },
    });
    expect(props.onChangeInterval).toBeCalledWith('w');
  });

  it('should match snapshot', () => {
    const { component } = renderTimechart();
    expect(component.container).toMatchSnapshot();
  });
});
