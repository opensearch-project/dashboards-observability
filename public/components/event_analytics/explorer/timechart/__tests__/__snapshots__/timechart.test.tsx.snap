// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timechart /> spec should match snapshot 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow dscResultCount"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiText euiText--medium"
      >
        <strong
          data-test-subj="discoverQueryHits"
        >
          194
        </strong>
         
        hits
      </div>
    </div>
  </div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <span
        class="euiToolTipAnchor"
      >
        <div
          class="euiText euiText--small"
          data-test-subj="discoverIntervalDateRange"
        >
          Jan 1, 2024 @ 00:00:00.000 - Jan 1, 2024 @ 00:00:00.000
        </div>
      </span>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFormControlLayout euiFormControlLayout--compressed"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <select
            aria-label="Time interval"
            class="euiSelect euiSelect--compressed"
            data-test-subj="eventAnalytics__EventIntervalSelect"
            id="dscResultsIntervalSelector"
          >
            <option
              value="m"
            >
              Minute
            </option>
            <option
              value="h"
            >
              Hour
            </option>
            <option
              value="d"
            >
              Day
            </option>
            <option
              value="w"
            >
              Week
            </option>
            <option
              value="M"
            >
              Month
            </option>
            <option
              value="y"
            >
              Year
            </option>
          </select>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--small euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157 8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0Z"
                  fill-rule="non-zero"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    id="explorerPlotComponent"
    style="width: 100%; height: 100%;"
  />
</div>
`;
